cltk/lemmatize/latin/backoff.py

```diff
@@ -30,433 +32,257 @@
 
 
-class NgramLemmatizer(ContextLemmatizer, NgramTagger):
-    """"""
-    def __init__(self, n, train=None, model=None, backoff=None, cutoff=0):
-        """Setup for NgramLemmatizer()
+class UnigramLemmatizer(SequentialBackoffLemmatizer, UnigramTagger):
+    """
+    Standalone version of 'train' function found in UnigramTagger; by
+    defining as its own class, it is clearer that this lemmatizer is
+    based on training data and not on dictionary.
+    """
 
-        :param n: Int with length of 'n'-gram
-        :param train: List of tuples of the form (TOKEN, LEMMA)
-        :param model: Dict; DEPRECATED, use TrainLemmatizer
-        :param backoff: Next lemmatizer in backoff chain.
-        :param cutoff: Int with minimum number of matches to choose lemma
+    def __init__(self, train=None, model=None, backoff=None, source=None, cutoff=0, VERBOSE=False):
         """
-        self._n = n
-        self._check_params(train, model)
-        ContextLemmatizer.__init__(self, model, backoff)
-        NgramTagger.__init__(self, self._n, train, model, backoff, cutoff)
-
-        if train:
-            # Refactor to remove model? Always train?
-            self._train(train, cutoff)
-
-    def context(self, tokens, index, history):
-        """"""
-        return NgramTagger.context(self, tokens, index, history)
+        Setup for UnigramLemmatizer()
+        """
+        SequentialBackoffLemmatizer.__init__(self, backoff=None, VERBOSE=VERBOSE)
+        UnigramTagger.__init__(self, train, model, backoff, cutoff)
+        self.train = train
+        self.source = source
 
 
-class UnigramLemmatizer(NgramLemmatizer, UnigramTagger):
-    """Setup for UnigramLemmatizer()"""
-    def __init__(self, train=None, model=None, backoff=None, cutoff=0):
-        """"""
-        NgramLemmatizer.__init__(self, 1, train, model, backoff, cutoff) # Note 1 for unigram
-        UnigramTagger.__init__(self, train, model, backoff, cutoff)
+    def __repr__(self):
+        if self.source:
+            return f'<UnigramLemmatizer: {self.source}>'
+        else:
+            return f'<UnigramLemmatizer: {self.repr.repr(self.train)}>'
 
 
 class RegexpLemmatizer(SequentialBackoffLemmatizer, RegexpTagger):
     """"""
 
-    def __init__(self, regexps=None, backoff=None):
+    def __init__(self, regexps=None, source=None, backoff=None, VERBOSE=False):
         """Setup for RegexpLemmatizer()
-
+        :type regexps: list
         :param regexps: List of tuples of form (PATTERN, REPLACEMENT)
         :param backoff: Next lemmatizer in backoff chain.
         """
-        SequentialBackoffLemmatizer.__init__(self, backoff)
+        SequentialBackoffLemmatizer.__init__(self, backoff=None, VERBOSE=VERBOSE)
         RegexpTagger.__init__(self, regexps, backoff)
         self._regexs = regexps
+        self.source = source
+
 
-    def choose_lemma(self, tokens, index, history):
+    def choose_tag(self, tokens, index, history):
```

Comment:
consider adding type annotations, since the param input types are well defined below

Reply:
Good point—still hasn't become part of my workflow, nice to have a reminder.